contracts/asset-proxy/contracts/src/bridges/KyberBridge.sol

```diff
@@ -0,0 +1,160 @@
+    IWallet
+{
+    // @dev Structure used internally to get around stack limits.
+    struct TradeState {
+        IKyberNetworkProxy kyber;
+        IEtherToken weth;
+        address fromTokenAddress;
+        uint256 fromTokenBalance;
+        uint256 payableAmount;
+        uint256 minConversionRate;
+    }
+
+    /// @dev Address of the WETH contract.
+    address constant public WETH_ADDRESS = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    /// @dev Address of the KyberNeworkProxy contract.
+    address constant public KYBER_NETWORK_PROXY_ADDRESS = 0x818E6FECD516Ecc3849DAf6845e3EC868087B755;
+    /// @dev Kyber ETH pseudo-address.
+    address constant public KYBER_ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /// @dev Callback for `IKyberBridge`. Tries to buy `amount` of
+    ///      `toTokenAddress` tokens by selling the entirety of the opposing asset
+    ///      to the `KyberNetworkProxy` contract, then transfers the bought
+    ///      tokens to `to`.
+    /// @param toTokenAddress The token to give to `to`.
+    /// @param to The recipient of the bought tokens.
+    /// @param amount Minimum amount of `toTokenAddress` tokens to buy.
+    /// @param bridgeData The abi-encoeded "from" token address.
+    /// @return success The magic bytes if successful.
+    function bridgeTransferFrom(
+        address toTokenAddress,
+        address /* from */,
+        address to,
+        uint256 amount,
+        bytes calldata bridgeData
+    )
+        external
+        returns (bytes4 success)
+    {
+        TradeState memory state;
+        state.kyber = _getKyberContract();
+        state.weth = _getWETHContract();
+        // Decode the bridge data to get the `fromTokenAddress`.
+        (state.fromTokenAddress) = abi.decode(bridgeData, (address));
+        state.fromTokenBalance = IERC20Token(state.fromTokenAddress).balanceOf(address(this));
+        if (state.fromTokenBalance == 0) {
+            // Do nothing if no input tokens.
+            return BRIDGE_SUCCESS;
+        }
+        state.minConversionRate = (10 ** 18) * amount / state.fromTokenBalance;
+        if (state.fromTokenAddress == toTokenAddress) {
+            // Just transfer the tokens if they're the same.
+            LibERC20Token.transfer(state.fromTokenAddress, to, state.fromTokenBalance);
+            return BRIDGE_SUCCESS;
+        } else if (state.fromTokenAddress != address(state.weth)) {
+            // If the input token is not WETH, grant an allowance to the exchange
+            // to spend them.
+            LibERC20Token.approve(state.fromTokenAddress, address(state.kyber), uint256(-1));
+        } else {
+            // If the input token is WETH, unwrap it and attach it to the call.
+            state.fromTokenAddress = KYBER_ETH_ADDRESS;
+            state.payableAmount = state.fromTokenBalance;
+            state.weth.withdraw(state.fromTokenBalance);
+        }
+
+        // Try to sell all of this contract's input token balance.
```

Comment:
Does `trade` always revert on failure? Is there a scenario where it just returns  an invalid bought amount, like zero?

Reply:
Should revert but I guess with a bad enough conversion rate it could return zero. The asset proxy should catch it on the balance check, if it's too low.